include/bls12-381/scalar.hpp

```diff
@@ -215,21 +215,39 @@ void rsh(std::array<uint64_t, N>& out, const std::array<uint64_t, N>& in, uint64
 
 void bn_divn_low(uint64_t *c, uint64_t *d, uint64_t *a, int sa, uint64_t *b, int sb);
 
+template<size_t N, size_t M>
+void bn_divn_safe(std::array<uint64_t, N>& c, std::array<uint64_t, M>& d, const std::array<uint64_t, N>& a, const std::array<uint64_t, M>& b) {
+    
+    static_assert(N >= M, "dividend must be at least same word size as modulus");
+
+    std::array<uint64_t, N+2> modulus = {0};
+    std::array<uint64_t, N+2> dividend = {0};
+
+    memcpy(dividend.data(), a.data(), N * sizeof(uint64_t));
```

Comment:
same as #5 solution this one seems to copy array as well. Is this necessary? Why can't we use original `k` array with bigger size?

Reply:
k is usually stored as const. But the algorithm may modify that and will even access beyond origin size. So we need a copy of k in a larger buffer.

Reply:
so if `k` is `const` and unused you could make it initially 2 bytes bigger and pass to `bn_divn_low` not `N` but `N-2`. That will avoid buffer reallocation every time you call this routine. Just need to make sure you use `modPrime<10>` instead of `modPrime<8>`, etc.

Reply:
this approach will make the upper level code a little bit less intuitive, i.e. always use 10 words buffer to hold 8 words data. Easy to introduce mistakes. And that's the reason why we choose to fix the buffer in the way this pr does instead of fix it in everywhere calling the bn_divn_low function.

Reply:
ok, I got you. what about this trick? That way you can set template parameter to 'N', avoid copy and deal with it like it has fewer elements. I'm not sure we need `operator=`, maybe you can make `padded_array` structure as a return type from `scalar::fromBytesBE` to avoid copying there.

```c++
#include <array>

using namespace std;

template<typename T, size_t Num>
struct padded_array : public array<T, Num+2> {
    using base = array<T, Num+2>;
    // this makes a copy but I'm not sure we are really need it.
    template<size_t Num2>
    padded_array<T, Num>& operator=( const array<T, Num2>& arr2 ) {
        copy(arr2.begin(), arr2.end(), base::begin());
        return *this;
    }
};

int main(int, char**) {
    padded_array<int, 8> arr = {1,2,3,4,5,6,7,8};
    array<int, 8> arr2 = {8,7,6,5,4,3,2,1};
    arr = arr2;

    return 0;
}
```

Reply:
Nice trick. In the `operator=()`, I would `static_assert` that `Num2 <= Num+2`.
This would definitely be a valid solution!
However it would still be more intrusive than the current solution (require modifications in all call sites), so if the current solution does not introduce a significant slowdown, I would still suggest to go with it and not attempt further optimizations.

Reply:
this approach would pretty much mean a special version of scalar::fromBytesBE

Seems a little bit overkill if the performance gain is not significant.. 

I will mark this as resolved. Feel free to unresolve it if necessary.